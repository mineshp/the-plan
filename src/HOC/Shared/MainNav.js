import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { logout } from '../../actions/authentication';
import MainNavComponent from '../../components/MainNav';

class MainNav extends Component {
    constructor(props) {
        super(props);
        this.logout = this.logout.bind(this);
    }

    logout() {
        this.props.actions.logout();
    }

    /* istanbul ignore next: not testing render */
    render() {
        return (
            <MainNavComponent
                username={this.props.username}
                isAdmin={this.props.isAdmin}
                logout={this.logout}
            />
        );
    }
}

/* istanbul ignore next: not testing mapStateToProps */
const mapStateToProps = (state) => ({
    username: state.authentication.user.username,
    isAdmin: state.authentication.user.isAdmin
});

/* istanbul ignore next: not testing mapDispatchToProps */
const mapDispatchToProps = (dispatch) => (
    {
        actions: bindActionCreators({ logout }, dispatch)
    }
);

const MainNavConnectedComponent = connect(
    mapStateToProps,
    mapDispatchToProps
)(MainNav);

MainNav.propTypes = {
    username: PropTypes.string,
    isAdmin: PropTypes.bool,
    actions: PropTypes.shape({
        logout: PropTypes.func.isRequired
    })
};

MainNav.defaultProps = {
    username: null,
    isAdmin: false,
    actions: null,
};

export {
    MainNav,
    MainNavConnectedComponent
};
