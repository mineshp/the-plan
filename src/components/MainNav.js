import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Dropdown, Menu } from 'semantic-ui-react';
import logo from '../md-logo-green.png';

export default class MainNav extends Component {
    constructor(props) {
        super(props);
        this.state = {
            activeItem: 'view',
            moto: null
        };
        this.handleItemClick = this.handleItemClick.bind(this);
        this.logout = this.logout.bind(this);
    }

    // eslint-disable-next-line class-methods-use-this
    getMotoByUser({ username }) {
        let moto;
        switch (username) {
        case 'minesh':
            moto = 'Haz tus sue√±os realidad';
            break;
        case 'Jignasha':
            moto = 'Te amo mucho, Bella';
            break;
        default:
            moto = 'Make your dreams a reality';
            break;
        }
        return moto;
    }

    handleItemClick(e, { name }) {
        this.setState({ activeItem: name });
    }

    logout(e) {
        e.preventDefault();
        this.props.logout();
    }

    render() {
        const { activeItem } = this.state;
        const { isAdmin, username } = this.props;
        const displayName = username
            ? `${username.charAt(0).toUpperCase() + username.slice(1)}`
            : null;
        const displayControlCenterItem = isAdmin
            ? <Dropdown.Item onClick={this.handleItemClick} href="/admin/manage">Control Center</Dropdown.Item>
            : null;
        const moto = this.getMotoByUser({ username });
        return (
            <div className="header-main">
                <Menu>
                    <Menu.Item
                        href="/"
                        name="home"
                        active={activeItem === 'home'}
                        onClick={this.handleItemClick}
                    >
                        <img alt="logo" src={logo} />
                    </Menu.Item>

                    <Dropdown item text="Projects">
                        <Dropdown.Menu>
                            <Dropdown.Item onClick={this.handleItemClick} href="/project/all">List</Dropdown.Item>
                            <Dropdown.Item onClick={this.handleItemClick} href="/project/update">Create</Dropdown.Item>
                            <Dropdown.Item disabled>Search</Dropdown.Item>
                        </Dropdown.Menu>
                    </Dropdown>

                    <Dropdown item text="Lists">
                        <Dropdown.Menu>
                            <Dropdown.Item onClick={this.handleItemClick} href="/list/all">List</Dropdown.Item>
                            <Dropdown.Item onClick={this.handleItemClick} href="/list/update">Create</Dropdown.Item>
                            <Dropdown.Item disabled>Search</Dropdown.Item>
                        </Dropdown.Menu>
                    </Dropdown>

                    <Menu.Item>
                        <span className="moto">{moto}</span>
                    </Menu.Item>

                    {
                        this.props.username
                            ?
                            (
                                <Menu.Item position="right">
                                    <Dropdown text={displayName} icon="user" floating labeled button className="icon blue">
                                        <Dropdown.Menu>
                                            {displayControlCenterItem}
                                            <Dropdown.Item onClick={this.logout}>Logout</Dropdown.Item>
                                        </Dropdown.Menu>
                                    </Dropdown>
                                </Menu.Item>
                            )
                            :
                            (<Menu.Item
                                href="/user/login"
                                position="right"
                                name="sign-in"
                                active={activeItem === 'sign-in'}
                                onClick={this.handleItemClick}
                            >Sign-in
                            </Menu.Item>)
                    }
                </Menu>
            </div>
        );
    }
}

MainNav.propTypes = {
    username: PropTypes.string,
    isAdmin: PropTypes.bool,
    logout: PropTypes.func.isRequired
};

MainNav.defaultProps = {
    username: null,
    isAdmin: false
};
