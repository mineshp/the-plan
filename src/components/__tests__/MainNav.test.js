import React from 'react';
import MainNav from '../MainNav';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { Menu } from 'semantic-ui-react'

describe('MainNav Component', () => {
    it('renders correctly', () => {
        const tree = renderer.create(<MainNav />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('ensure handleclick fires', () => {
        const mockHandleClick = jest.fn();

        const wrapper = shallow(<MainNav />);
        const firstMenuItem = wrapper.find(Menu.Item).first();
        firstMenuItem.props().onClick(null, { name: 'home' });

        expect(wrapper.state().activeItem).toBe('home');
    });
});