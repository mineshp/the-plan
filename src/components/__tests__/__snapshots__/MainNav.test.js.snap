// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainNav Component renders correctly 1`] = `
<div
  className="ui menu"
>
  <a
    className="item"
    href="/"
    onClick={[Function]}
  >
    <img
      alt="logo"
      src="md-logo.png"
    />
  </a>
  <a
    className="item"
    href="/list"
    onClick={[Function]}
  >
    List
  </a>
  <div
    aria-busy={undefined}
    aria-disabled={undefined}
    aria-expanded={false}
    aria-multiselectable={undefined}
    className="ui item dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    role="listbox"
    tabIndex={0}
  >
    <div
      aria-live="polite"
      className="text"
      role="alert"
    >
      Projects
    </div>
    <i
      aria-hidden="true"
      className="dropdown icon"
      onClick={[Function]}
    />
    <div
      className="menu transition"
    >
      <a
        aria-checked={undefined}
        aria-disabled={undefined}
        aria-selected={undefined}
        className="item"
        href="/project/all"
        onClick={[Function]}
        role="option"
      >
        List
      </a>
      <a
        aria-checked={undefined}
        aria-disabled={undefined}
        aria-selected={undefined}
        className="item"
        href="/project/update"
        onClick={[Function]}
        role="option"
      >
        Create
      </a>
      <div
        aria-checked={undefined}
        aria-disabled={undefined}
        aria-selected={undefined}
        className="item"
        onClick={[Function]}
        role="option"
      >
        Search
      </div>
    </div>
  </div>
  <div
    aria-busy={undefined}
    aria-disabled={undefined}
    aria-expanded={false}
    aria-multiselectable={undefined}
    className="ui item dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    role="listbox"
    tabIndex={0}
  >
    <div
      aria-live="polite"
      className="text"
      role="alert"
    >
      Lists
    </div>
    <i
      aria-hidden="true"
      className="dropdown icon"
      onClick={[Function]}
    />
    <div
      className="menu transition"
    >
      <a
        aria-checked={undefined}
        aria-disabled={undefined}
        aria-selected={undefined}
        className="item"
        href="/list/all"
        onClick={[Function]}
        role="option"
      >
        List
      </a>
      <a
        aria-checked={undefined}
        aria-disabled={undefined}
        aria-selected={undefined}
        className="item"
        href="/list/update"
        onClick={[Function]}
        role="option"
      >
        Create
      </a>
      <div
        aria-checked={undefined}
        aria-disabled={undefined}
        aria-selected={undefined}
        className="item"
        onClick={[Function]}
        role="option"
      >
        Search
      </div>
    </div>
  </div>
  <a
    className="right item"
    onClick={[Function]}
  >
    Sign-in
  </a>
</div>
`;
